core/modules/widgets/data.js

```diff
@@ -0,0 +1,144 @@
+
+var DataWidget = function(parseTreeNode,options) {
+	this.dataWidgetTag = parseTreeNode.type;
+	this.initialise(parseTreeNode,options);
+};
+
+/*
+Inherit from the base widget class
+*/
+DataWidget.prototype = new Widget();
+
+/*
+Render this widget into the DOM
+*/
+DataWidget.prototype.render = function(parent,nextSibling) {
+	this.parentDomNode = parent;
+	this.computeAttributes();
+	this.execute();
+	var domNode = this.document.createTextNode("");
+	parent.insertBefore(domNode,nextSibling);
+	this.renderChildren(parent,nextSibling);
+	// Children must have been rendered before we can read the data values
+	domNode.textContent = JSON.stringify(this.readDataTiddlerValues());
+	this.domNodes.push(domNode);
+};
+
+/*
+Compute the internal state of the widget
+*/
+DataWidget.prototype.execute = function() {
+	// Construct the child widgets
+	this.makeChildWidgets();
+};
+
+/*
+Read the tiddler value(s) from a data widget – must be called after the .render() method
+*/
+DataWidget.prototype.readDataTiddlerValues = function() {
+	var self = this;
+	// Start with a blank object
+	var item = Object.create(null);
+	// Read any attributes not prefixed with $
+	$tw.utils.each(this.attributes,function(value,name) {
+		if(name.charAt(0) !== "$") {
+			item[name] = value;	
+		}
+	});
+	item = new $tw.Tiddler(item);
+	// Deal with $tiddler, $filter or $compound-tiddler attributes
+	var tiddlers = [],title;
+	if(this.hasAttribute("$tiddler")) {
+		title = this.getAttribute("$tiddler");
+		if(title) {
+			tiddlers.push(this.wiki.getTiddler(title));
+		}
+	}
+	if(this.hasAttribute("$filter")) {
+		var filter = this.getAttribute("$filter");
+		if(filter) {
+			var titles = this.wiki.filterTiddlers(filter);
+			$tw.utils.each(titles,function(title) {
+				var tiddler = self.wiki.getTiddler(title);
+				tiddlers.push(tiddler);
+			});
+		}
+	}
+	if(this.hasAttribute("$compound-tiddler")) {
+		title = this.getAttribute("$compound-tiddler");
+		if(title) {
+			tiddlers.push.apply(tiddlers,this.extractCompoundTiddler(title));
+		}
+	}
+	// Convert the literal item to field strings
+	item = item.getFieldStrings();
+	if(tiddlers.length === 0) {
+		if(Object.keys(item).length > 0 && !!item.title) {
+			return [item];
+		} else {
+			return [];
+		}
+	} else {
+		var results = [];
+		$tw.utils.each(tiddlers,function(tiddler,index) {
+			var fields = tiddler.getFieldStrings();
+			results.push($tw.utils.extend({},fields,item));
+		});
+		return results;	
+	}
+};
+
+/*
+Helper to extract tiddlers from text/vnd.tiddlywiki-multiple tiddlers
+*/
+DataWidget.prototype.extractCompoundTiddler = function(title) {
+	var tiddler = this.wiki.getTiddler(title);
+	if(tiddler && tiddler.fields.type === "text/vnd.tiddlywiki-multiple") {
+		var text = tiddler.fields.text || "",
+			rawTiddlers = text.split("\n+\n"),
+			tiddlers = [];
+		$tw.utils.each(rawTiddlers,function(rawTiddler) {
+			var fields = Object.create(null),
+				split = rawTiddler.split(/\r?\n\r?\n/mg);
```

Comment:
`rawTiddlers` are split with 2 or more NLs ... `text.split("\n+\n")` 
can `rawTiddler` be split with 2 new-lines again `rawTiddler.split(/\r?\n\r?\n/mg);` ... Did I misread something here?

Reply:
Thanks @pmario – `text.split("\n+\n")` is not a regular expression, it's matching a plus sign surrounded by newlines. I think you're right that we should include support for carriage returns.